### ISSUE TITLE
Duplication of dynamic block's container element with API v2 in Gutenberg 10.2.0

### ISSUE THREAD

**Issue by rodrigodagostino** [WordPress/gutenberg#29953]
## Description
Right after upgrading to Gutenberg 10.2.0, some of my custom blocks (working with the API v2 and the `useBlockProps()` function) have now a double block container in the editor, like the following:

```html
<div id="block-d0344daf-eb2e-467f-8905-5084c3783ff7" tabindex="0" role="group" aria-label="Block: Weather"
    data-block="d0344daf-eb2e-467f-8905-5084c3783ff7" data-type="custom-blocks/weather" data-title="Weather"
    class="wp-block block-editor-block-list__block">

    <div class="wp-block custom-block-weather block-editor-block-list__block"
        id="block-d0344daf-eb2e-467f-8905-5084c3783ff7" tabindex="0" role="group" aria-label="Block: Weather"
        data-block="d0344daf-eb2e-467f-8905-5084c3783ff7" data-type="custom-blocks/weather" data-title="Weather">
```

I’ve tested it on Gutenberg 10.2.0 RC1 and in WP 5.7.0 (Gutenberg deactivated), and this behaviour is also present there.

This issue was not present on Gutenberg 10.1.0/10.1.1.

The front-end side of the matter is working as expected, with only one container element and without any kind of issues.

## Code snippet
index.js
```javascript
/**
 * BLOCK: Weather
 *
 * Allows the user to display the current weather conditions for a chosen location.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n'
import { InnerBlocks } from '@wordpress/block-editor'

/**
 * Internal dependencies
 */
import edit from './edit'
import icon from './icon'
import metadata from './block.json'


const { name } = metadata

export { metadata, name }

export const settings = {
    title: __( 'Weather' ),
    description: __( 'Display the weather conditions for a chosen location.' ),
    icon,
    keywords: [
        /* translators: block keyword */
        __( 'Weather' ),
        /* translators: block keyword */
        __( 'custom block' ),
    ],

    edit,
    save: () => {
        return <InnerBlocks.Content />
    },
}


// Provide a custom block class
function setBlockCustomClassName( className, blockName ) {
    return blockName === name ?
        'custom-block-weather' :
        className
}

wp.hooks.addFilter(
    'blocks.getBlockDefaultClassName',
    'custom-blocks/custom-block-weather',
    setBlockCustomClassName,
)
```

edit.js
```javascript
/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n'
import { getBlockDefaultClassName } from '@wordpress/blocks'
import { InspectorControls, useBlockProps } from '@wordpress/block-editor'
import { PanelBody, Placeholder, SelectControl, TextControl, ToggleControl } from '@wordpress/components'
import { useEffect, useRef, useState } from '@wordpress/element'

/**
 * Internal dependencies
 */
import metadata from './block.json'
import { languages } from './components/definitions'
import LocationSelectControl from './components/location-select-control'
import Spinner from '../../components/spinner'

const { name } = metadata

function WeatherEdit( {
    className,
    attributes,
    attributes: { cityName, shortName, displayUnits, units, language },
    setAttributes,
} ) {
    const [ weatherData, setWeatherData ] = useState( null )
    const [ isFetching, setIsFetching ] = useState( true )

    const prevCityName = useRef( 'Río Grande,AR' )
    const prevUnits = useRef( 'metric' )

    useEffect( () => {
        if ( ! language ) {
            const lang = pluginData.locale.substr( 0, 2 )
            setAttributes( { language: lang } )
        }

        fetchWeatherData()
    }, [] )

    useEffect( () => {
        if ( prevCityName.current !== cityName || prevUnits.current !== units ) {
            fetchWeatherData()
        }
    }, [ cityName, units ] )

    const fetchWeatherData = async () => {
        setIsFetching( true )

        const url = `https://api.openweathermap.org/data/2.5/weather?q=${ cityName }&appid=${ pluginData.settings.openWeatherApiKey }&units=${ units }&lang=${ language }`

        await fetch( url )
            .then( ( results ) => {
                return results.json()
            } )
            .then( ( data ) => {
                setWeatherData( data )
                setIsFetching( false )
            } )
            .catch( ( error ) => console.error( error ) )
    }

    const defaultClassName = getBlockDefaultClassName( name )

    const blockProps = useBlockProps( { className } )

    let dayNight = ''
    if ( weatherData ) {
        const timeStamp = weatherData.dt
        const sunrise = weatherData.sys.sunrise
        const sunset = weatherData.sys.sunset
        if ( timeStamp > sunrise && timeStamp < sunset ) {
            dayNight = 'day'
        } else {
            dayNight = 'night'
        }
    }

    return (
        <>
            <InspectorControls>
                <PanelBody title={ __( 'Main' ) }>
                    <LocationSelectControl { ...{ attributes, setAttributes } } />

                    <TextControl
                        label={ __( 'Short name' ) }
                        value={ shortName }
                        onChange={ ( value ) => setAttributes( { shortName: value } ) }
                    />

                    <ToggleControl
                        label={ __( 'Display units' ) }
                        checked={ displayUnits }
                        onChange={ () => setAttributes( { displayUnits: ! displayUnits } ) }
                    />

                    <SelectControl
                        label={ __( 'Units' ) }
                        labelPosition="top"
                        value={ units }
                        options={ [
                            { label: __( 'Metric' ), value: 'metric' },
                            { label: __( 'Imperial' ), value: 'imperial' },
                            { label: __( 'Kelvin' ), value: 'kelvin' },
                        ] }
                        onChange={ ( value ) => setAttributes( { units: value } ) }
                    />

                    <SelectControl
                        label={ __( 'Language' ) }
                        labelPosition="top"
                        value={ language }
                        options={ languages }
                        onChange={ ( value ) => setAttributes( { language: value } ) }
                    />
                </PanelBody>
            </InspectorControls>

            <div { ...blockProps }>
                { !! isFetching && (
                    <Placeholder>
                        <Spinner />
                        <p>{ __( 'Loading…' ) }</p>
                    </Placeholder>
                ) }

                { ! isFetching && !! weatherData && (
                    <>
                        { !! shortName && (
                            <style>
                                { `.${ defaultClassName }__city-name { display: none }
                            @media only screen and (min-width: ${ pluginData.settings.xlargeBreakpointValue }px) {
                                .${ defaultClassName }__short-name { display: none }
                                .${ defaultClassName }__city-name { display: inline-block }
                            }` }
                            </style>
                        ) }

                        <i
                            className={ `${ defaultClassName }__icon wi wi-owm-${ dayNight }-${ weatherData.weather[0].id }` }
                            alt={ weatherData.description }
                        ></i>
                        <div className={ `${ defaultClassName }__temperature` }>
                            <span className={ `${ defaultClassName }__temperature-number` }>
                                { Math.round( weatherData.main.temp ) }
                            </span>
                            { ( units === 'metric' || units === 'imperial' ) && (
                                <span className={ `${ defaultClassName }__temperature-degree-symbol` }>&deg;</span>
                            ) }
                            { displayUnits && units === 'metric' && (
                                <span className={ `${ defaultClassName }__temperature-units` }>C</span>
                            ) }
                            { displayUnits && units === 'imperial' && (
                                <span className={ `${ defaultClassName }__temperature-units` }>F</span>
                            ) }
                            { displayUnits && units === 'kelvin' && (
                                <span className={ `${ defaultClassName }__temperature-units` }>K</span>
                            ) }
                        </div>
                        <span className={ `${ defaultClassName }__short-name` }>{ shortName }</span>
                        <span className={ `${ defaultClassName }__city-name` }>{ weatherData.name }</span>
                    </>
                ) }
            </div>
        </>
    )
}

export default WeatherEdit
```

## Expected behaviour
Blocks being rendered within only one container element.

## Actual behaviour
Blocks being rendered within two container elements.

## WordPress information
- WordPress version: 5.7.0
- Gutenberg version: 10.2.0
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme? Yes

## Device information
- Device: Desktop
- Operating system: Windows 10
- Browser: Firefox Developer 87.0b9, Chrome 89.0.4389.82

**Comment by talldan** [comment]
@rodrigodagostino Would you be able to provide a code sample for an affected block?

**Comment by rodrigodagostino** [comment]
Sure thing, @talldan . I apologize for not doing it beforehand. I've just updated my initial comment to include the proper code snippets. Hope that helps clarifying the situation.

**Comment by rodrigodagostino** [comment]
What is the actual expected output? Maybe I'm trying to fight a double wrapper that should be there, but WP core blocks do have them.

**Comment by thomasplevy** [comment]
@rodrigodagostino Try declaring `apiVersion: 2` when registering your block:

```js
export const settings = {
   apiVersion: 2,
   // the rest of your settings.
}
```

I was having this problem until finding the apiVersion on an example doc and then finding this blog post: https://make.wordpress.org/core/2020/11/18/block-api-version-2/

**Comment by rodrigodagostino** [comment]
@thomasplevy thank you very much for your contribution, but that was there since my first attempt :(

The goods news is (after more digging) I realised that this issue was only happening with my dynamic blocks. And after even more digging into the core blocks code, I found out that the way to solve it was to replace the `register_block_type()` with `register_block_type_from_metadata()` to register my block on the server side.

Don't really know why, but that did the trick, and the appearance of my blocks is now as expected :) I tested this on Gutenberg v10.3.1, and even on WP v5.7 with the Gutenberg plugin deactivated.

There's no info around about API v2 and dynamic blocks, so I hope this helps someone out there. I will adjust the issue title to help others find it.

And I guess now we can close this issue :)
